File: algorithms/sorting/heap-sort/HeapSortTest.java
import java.util.Arrays;
import java.util.Random;

public final class HeapSortTest {
    private HeapSortTest() {
    }

    private static void verify(String name, int[] input, int[] expected) {
        int[] original = Arrays.copyOf(input, input.length);
        int[] actual = HeapSort.heapSort(input);
        if (!Arrays.equals(actual, expected)) throw new AssertionError(name + " failed: " + Arrays.toString(actual));
        if (!Arrays.equals(input, original)) throw new AssertionError(name + " mutated its input");
    }

    public static void main(String[] arguments) {
        verify("empty input", new int[] {}, new int[] {});
        verify("single value", new int[] {7}, new int[] {7});
        verify("already sorted", new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5});
        verify("reverse order", new int[] {5, 4, 3, 2, 1}, new int[] {1, 2, 3, 4, 5});
        verify("duplicates", new int[] {4, 2, 4, 1}, new int[] {1, 2, 4, 4});
        verify("negative values", new int[] {-2, 5, -8, 0}, new int[] {-8, -2, 0, 5});

        Random random = new Random(83);
        for (int round = 0; round < 30; round++) {
            int[] input = new int[12];
            for (int index = 0; index < input.length; index++) input[index] = random.nextInt(101) - 50;
            int[] expected = Arrays.copyOf(input, input.length);
            Arrays.sort(expected);
            verify("random input " + round, input, expected);
        }

        System.out.println("Java Heap Sort tests passed");
    }
}

File: tests/heap-sort.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { heapSort } from "../algorithms/sorting/heap-sort/heap-sort.mjs";

const cases = [
  { name: "empty input", input: [], expected: [] },
  { name: "single value", input: [7], expected: [7] },
  { name: "already sorted", input: [1, 2, 3, 4, 5], expected: [1, 2, 3, 4, 5] },
  { name: "reverse order", input: [5, 4, 3, 2, 1], expected: [1, 2, 3, 4, 5] },
  { name: "duplicates", input: [4, 2, 4, 1], expected: [1, 2, 4, 4] },
  { name: "negative values", input: [-2, 5, -8, 0], expected: [-8, -2, 0, 5] },
];

for (const { name, input, expected } of cases) {
  test(`Heap Sort handles ${name}`, () => {
    const original = [...input];
    assert.deepEqual(heapSort(input), expected);
    assert.deepEqual(input, original, "input should not be mutated");
  });
}

test("Heap Sort agrees with the platform sort oracle for deterministic random inputs", () => {
  let seed = 83;
  const next = () => {
    seed = (seed * 48271) % 2147483647;
    return seed;
  };

  for (let round = 0; round < 30; round += 1) {
    const input = Array.from({ length: 12 }, () => (next() % 101) - 50);
    assert.deepEqual(heapSort(input), [...input].sort((a, b) => a - b));
  }
});
